### DISCUSSION TITLE
Import raw  image data without OpenCV

### DISCUSSION THREAD

**Discussion by kalwalt** [davisking/dlib#2687] [Q&A]
Hi to all! Thank for this great and nice project!
I am Walter Perdan @kalwalt with @ThorstenBux i am co-founder of [WebARKit](https://github.com/webarkit) a small github organization for making WebAR (Augmented Reality for the Web).
I'm developing a little project with dlib [dlib_em.js](https://github.com/kalwalt/dlib_em.js) and Emscripten. The main idea is to provide some base code to interface dlib to be transpiled with Emscripten. For now there is only one example for loading a simple image, apply some transformation and display the result, this is the simple code:

```c++
#include "include/image_example.h"


/* This example was taken by dlib exmples https://github.com/davisking/dlib/blob/master/examples/image_ex.cpp 
   For more informations refer to comments in that. 
   In this case we simply load a jpeg image and apply a gaussian filter and a sobel edge detector.
   The result of image processing is returned outside the function. See transformImage in dlib_em.cpp how is handled.
*/
array2d<unsigned char> transform_image(const std::string &file) {
  array2d<rgb_pixel> img;
  load_image(img, file);
  array2d<unsigned char> blurred_img;
  gaussian_blur(img, blurred_img);


  // Now find the horizontal and vertical gradient images.
  array2d<short> horz_gradient, vert_gradient;
  array2d<unsigned char> edge_image;
  sobel_edge_detector(blurred_img, horz_gradient, vert_gradient);


  // now we do the non-maximum edge suppression step so that our edges are nice
  // and thin
  suppress_non_maximum_edges(horz_gradient, vert_gradient, edge_image);
  return edge_image;
};
```
[here](https://github.com/kalwalt/dlib_em.js/blob/cd10151414282a78b57216ac8c9c0d8cc0554646/src/image_example.cpp#L1-L23) the original code.
I saw that in the source code there are some utilities to load image data to `cv::Mat` and viceversa but i would avoid using OpenCV in my project.
With emscripten it's possible to load raw data and pass to a C++ `vector` container in this way:

```C++
void load_data(emscripten::val data_buffer, size_t size) {
      std::vector<unsigned char> u8(size);
      u8 = emscripten::convertJSArrayToNumberVector<u_char>(data_buffer);
      // do someting with u8 data...
}

```
There is some utility to fill an array2d with a `std::vector`? or what do you suggest to do?
Thank you for your attention,
Walter alias @kalwalt.

**Comment by arrufat** [comment]
Maybe you could try with [`image_data`](http://dlib.net/dlib/image_processing/generic_image.h.html). It returns a `void*` to the raw image data.

**Reply by kalwalt** [reply]
@arrufat  thank you for the answer, I think i will try to mimick `cv_image `for this pourpose. I think it came close to what i need.
